File: src/main/java/io/github/Chase22/telegram/telegrambotapimock/infrastructure/server/handler/HandlerChain.java
package io.github.Chase22.telegram.telegrambotapimock.infrastructure.server.handler;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.github.Chase22.telegram.telegrambotapimock.config.TelegramBotApiConfiguration;
import io.github.Chase22.telegram.telegrambotapimock.infrastructure.server.Router;
import io.github.Chase22.telegram.telegrambotapimock.infrastructure.server.endpoint.EndpointRegistry;
import io.undertow.server.HttpHandler;
import io.undertow.server.handlers.BlockingHandler;
import io.undertow.server.handlers.PathTemplateHandler;

public class HandlerChain {
    private final HttpHandler handler;

    public HandlerChain(final EndpointRegistry endpointRegistry,
                        final TelegramBotApiConfiguration config,
                        final ObjectMapper objectMapper
    ) {

        ParameterHandler parameterHandler = new ParameterHandler(objectMapper);

        Router router = new Router(parameterHandler, endpointRegistry);
        BlockingHandler blockingHandler = new BlockingHandler(router);

        TokenHandler tokenHandler = new TokenHandler(config.getBotToken(), blockingHandler);

        PathTemplateHandler pathTemplateHandler = new PathTemplateHandler();
        pathTemplateHandler.add("/{id}/*", tokenHandler);

        handler = pathTemplateHandler;
    }

    public HttpHandler getHandler() {
        return handler;
    }
}

File: src/main/kotlin/io/github/chase22/telegram/telegrambotapimock/infrastructure/server/handler/HandlerChain.kt
package io.github.chase22.telegram.telegrambotapimock.infrastructure.server.handler

import com.fasterxml.jackson.databind.ObjectMapper
import io.github.chase22.telegram.telegrambotapimock.api.config.TelegramBotApiConfiguration
import io.github.chase22.telegram.telegrambotapimock.infrastructure.registry.EndpointRegistry
import io.undertow.server.HttpHandler
import io.undertow.server.handlers.GracefulShutdownHandler
import io.undertow.server.handlers.PathTemplateHandler
import io.undertow.server.handlers.RequestDumpingHandler

class HandlerChain(endpointRegistry: EndpointRegistry,
                   config: TelegramBotApiConfiguration,
                   objectMapper: ObjectMapper
) {
    val handler: HttpHandler

    init {
        val parameterHandler = ParameterHandler(objectMapper)

        val shutdownHandler = GracefulShutdownHandler(parameterHandler)

        val router = Router(shutdownHandler, endpointRegistry, objectMapper)

        val tokenHandler = TokenHandler(config.botToken, router)

        val pathTemplateHandler = PathTemplateHandler()
        pathTemplateHandler.add("/{id}/*", tokenHandler)

        val telegramErrorResponseHandler = TelegramErrorResponseHandler(pathTemplateHandler, objectMapper)

        handler = if (config.dumpRequests) {
            val requestDumpingHandler = RequestDumpingHandler(telegramErrorResponseHandler)
            requestDumpingHandler
        } else {
            telegramErrorResponseHandler
        }

    }
}
